### PR TITLE
Data Encryption Enhancement

### PR DESCRIPTION
**Pull request by developervariety**
#73

Proposes encryption for data, **does not encrypt metadata or any header fields.** Easily works with (no) compression, (no) SSL.

Default encryption algorithms:
- Aes
- TripleDes

`EncryptionPassphrase` must be 32 bytes or longer. A random salt is generated for every packet sent with encryption.

~`CustomEncryption` is used when `Encryption` is set to `EncryptionType.Custom`. The consumer must implement `IEncryption` into their custom encryption algorithm.~

### CODE REVIEW

**Comment by developervariety** [comment]
Sorry for the title changes, couldn't settle for any of them 😂.

Alright so this is a basic draft of how imagine it (w/ some improvements ofc). @jchristn please take a look at it when you get the chance, thank you!

File: WatsonTcp/WatsonMessage.cs [hunk 1]
```diff
@@ -106,6 +106,11 @@ public byte[] PresharedKey
         /// The type of compression used in the message.
         /// </summary>
         public CompressionType Compression = CompressionType.None;
+        
+        /// <summary>
+        /// The type of encryption used in the message.
+        /// </summary>
+        public EncryptionType Encryption = EncryptionType.Aes;
```
Comment (jchristn) [inline WatsonTcp/WatsonMessage.cs]: Should be EncryptionType.None by default

**Comment by jchristn** [comment]
Thanks @developervariety I will be taking a look at these over the next couple of days!

**Comment by developervariety** [comment]
No problem @jchristn! I'm going over stuff I missed and fixing any bugs I come across. Let me know when you're ready to add those header fields for encryption-related information. I don't want to get deep into the framing system!

**Comment by developervariety** [comment]
# # Current to-do list

- New header for encryption related information (aka `salt`, `algorithm`)
- Custom encryption extensibility
- Keep `ContentLength` in tact

**Comment by jchristn** [comment]
Hi @developervariety just a heads-up I may refactor WatsonMessage but it shouldn't impact what you're doing here - I'll likely move the compression bits out of WatsonMessage itself and try to move it into client and server.  Will keep you posted.

**Comment by developervariety** [comment]
Not a problem, I'll be sure to add any `master` branch changes once the PR is ready to be merged.

**Comment by developervariety** [comment]
Alright so I have updated to the latest WatsonTcp version and I'm still experiencing the `ContentLength` issue with an encrypted payload. I still have yet to decrypt received streams, just focusing on messages then I'll move onto that.

Perhaps I need a second pair of eyes.. not sure how I'm still having issues with encryption although the **entire data stream** is being read due to the new changes.

**Comment by jchristn** [comment]
Hi Justin I will check it out tomorrow!  Cheers sir and have a good night

**Comment by developervariety** [comment]
I think v4.1.4 should do the fix, I'll be merging the changes throughout the week when I get the time. Thank you for the time and dedication you put into this project boss, I really appreciate it.

**Comment by developervariety** [comment]
Hey @jchristn, I'm going to be updating the changes to `v4.1.5` soon. Is there anything that could possibly changed as it stands currently?

**Comment by jchristn** [comment]
Hi @developervariety I can't think of any changes in the queue that could affect you going forward.  I haven't looked too deeply into your PR to see how it's built out though.  Cheers, Joel

File: WatsonTcp/EncryptionHelper.cs [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+                }
+
+                return ms.ToArray();
+            }
+        }
+
+        public static byte[] Decrypt<T>(byte[] data, byte[] key, byte[] salt)
+            where T : SymmetricAlgorithm, new()
+        {
+            T algorithm = new T();
```
Comment (Job79) [inline WatsonTcp/EncryptionHelper.cs:33]: Add `using` to algorithm, now you are creating a memory leak.

File: WatsonTcp/WatsonCommon.cs [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+
+        internal static async Task<byte[]> ReadFromStreamAsync(Stream stream, long count, int bufferLen)
+        {
+            if (count <= 0) return null;
+            if (bufferLen <= 0) throw new ArgumentException("Buffer must be greater than zero bytes.");
+            byte[] buffer = new byte[bufferLen];
+
+            int read = 0;
+            long bytesRemaining = count;
+            MemoryStream ms = new MemoryStream();
```
Comment (Job79) [inline WatsonTcp/WatsonCommon.cs:74]: Here is another memory leak

Comment (jchristn) [reply WatsonTcp/WatsonCommon.cs:74]: This is intentional. The idea is to allow the consumer to close and dispose it.

**Comment by jchristn** [comment]
Misspoke, please ignore.

File: WatsonTcp/WatsonCommon.cs (hunk 3 above)
Comment (jchristn) [reply WatsonTcp/WatsonCommon.cs:74]: I fixed the leak in WatsonCommon.cs with NuGet v4.1.12.3: https://www.nuget.org/packages/WatsonTcp/4.1.12.3

The other one is in the PR so I'll leave that to @developervariety

Comment (Job79) [reply WatsonTcp/WatsonCommon.cs:74]: > I fixed the leak in WatsonCommon.cs with NuGet v4.1.12.3: https://www.nuget.org/packages/WatsonTcp/4.1.12.3
> 
> The other one is in the PR so I'll leave that to @developervariety

Thats a quick fix :+1: well done

**Comment by Job79** [comment]
> #73
> [~]
> `EncryptionPassphrase` must be 32 bytes or longer. A random salt is generated for every packet sent with encryption.


Currently the code is not generating a random `salt` every time. 
I don't think you want a random salt, this would change the key every time you send a message.
What you probably meant is an `IV`. This changes the output of the encrypt function every time it is used.
This is used to make it impossible for an attacker to detect routines in the traffic. (Every message is different, even if it has the same content)

Here are some examples:
```cs
// Encrypt
// Algorithm is an instance of a SymmetricAlgorithm with a key (Add key like this: Algorithm.Key = Key)
using var ms = new MemoryStream();
Algorithm.GenerateIV();
ms.Write(Algorithm.IV, 0, Algorithm.IV.Length); // Prepend IV to message

using var cs = new CryptoStream(ms, Algorithm.CreateEncryptor(Algorithm.Key, Algorithm.IV), CryptoStreamMode.Write);
cs.Write(data, 0, data.Length);
cs.FlushFinalBlock();

return ms.ToArray();
```
```cs
// Decrypt
byte[] iv = new byte[Algorithm.IV.Length];
Buffer.BlockCopy(data, 0, iv, 0, iv.Length); // Get (prepended) iv from data

Algorithm.IV = iv;
using var ms = new MemoryStream();
using var cs = new CryptoStream(ms, Algorithm.CreateDecryptor(Algorithm.Key, Algorithm.IV), CryptoStreamMode.Write);
cs.Write(data, iv.Length, data.Length - iv.Length); // Write data (without IV) to stream
cs.FlushFinalBlock();

return ms.ToArray();
```

**Comment by developervariety** [comment]
I personally do not have the time for this PR. If someone would like to continue with the draft please fork it off my account submit the PR here!

Thank you for understanding.

**Comment by jchristn** [comment]
Understood @developervariety thanks for letting us know.  A good alternative would be to use the library with SSL certificates.  Cheers, Joel

**Pull request closed without merging**